MCUX UART: Add support for HW flow control

mike-scott opened the pull request:
This patch series adds a DTS binding for 'hw-flow-control' and updates the MCUX uart shim to enable support for it.

Also, enable hw-flow-control in the modem overlay for lwm2m_client.

mike-scott committed:
dts: bindings: nxp-uart: add hw-flow-control binding

Support for HW flow control was recently enabled in the NXP MCUX
HAL layer.  Let's add a DTS binding definition for "hw-flow-control".
The MCUX uart driver shim can use this to enable HW flow control.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: serial: mcux: Introduce support for HW flow control

Now that a hw-flow-control DTS binding has been added for MCUX uart,
let's check for the DT_ define and enable support in the MCUX HAL
layer.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: lwm2m_client: enable HW flow control for WNC-M14A2A modem

The WNC-M14A2A modem is supposed to use HW flow control on the UART
connected to the MCU.  In the past, we were able to bypass this
requirement by using only the LTE-M firmware which supported pulling
the CTS line high via "send-ok" gpio.

This does NOT work for the LTE firmware.

Now that MCUX UART driver supports hw-flow-control setting, let's use
it in the modem overlay so that regardless of modem firmware
(LTE vs. LTE-M) the sample will "just work".

Signed-off-by: Michael Scott <mike@foundries.io>

galak commented on drivers/serial/uart_mcux.c:
```diff
@@ -289,6 +294,9 @@ static const struct uart_mcux_config uart_mcux_0_config = {
 	.clock_name = DT_UART_MCUX_0_CLOCK_NAME,
 	.clock_subsys = (clock_control_subsys_t)DT_UART_MCUX_0_CLOCK_SUBSYS,
 	.baud_rate = DT_UART_MCUX_0_BAUD_RATE,
+#if DT_UART_MCUX_0_HW_FLOW_CONTROL
```
Its pedantic, but I think its a little clearer to do:

`#if DT_UART_MCUX_0_HW_FLOW_CONTROL == 1`

mike-scott replied:
Yep, I'll change.

galak reviewed (commented):
Looks good, one comment on style of #if check.

mike-scott commented:
Updated #if checks to specifically check for a value of 1:
`#if DT_UART_MCUX_5_HW_FLOW_CONTROL`
to
`#if DT_UART_MCUX_5_HW_FLOW_CONTROL == 1`

galak commented on drivers/serial/uart_mcux.c:
```diff
@@ -289,6 +294,9 @@ static const struct uart_mcux_config uart_mcux_0_config = {
 	.clock_name = DT_UART_MCUX_0_CLOCK_NAME,
 	.clock_subsys = (clock_control_subsys_t)DT_UART_MCUX_0_CLOCK_SUBSYS,
 	.baud_rate = DT_UART_MCUX_0_BAUD_RATE,
+#if DT_UART_MCUX_0_HW_FLOW_CONTROL == 1
```
Actually why do you need the #if at all?

seems like:

`.hw_flow_control = DT_UART_MCUX_0_HW_FLOW_CONTROL,`

is fine since you'll get 0 or 1 depending on the property existing or not.

galak commented on samples/net/lwm2m_client/frdm_k64f.overlay:
```diff
@@ -5,6 +5,7 @@
  */
 &uart2 {
 	current-speed = <115200>;
+	hw-flow-control = <1>;
```
This should just be:

`hw-flow-control`;

Boolean are if the property name exists or not.

mike-scott commented:
Fixed the DTS change:
`hw-flow-control = <1>;` to `hw-flow-control;`

Removed the #if checks as the define will always be present for boolean properties.
